### ISSUE TITLE
FoldAndOrOfICmpsUsingRanges makes it hard to produce code compliant with BPF verifier

### ISSUE THREAD

**Issue by Spikhalskiy** [llvm/llvm-project#96847]
Currently when compiling code with BPF target, llvm makes optimizations that hurt further BPF verification and at the moment there are seemingly no real workarounds to this problem.

The problem is demonstrated by the following example:

Source code

```c
id = ctx->protocol;
if (id < 4 || id > 12)
 return 0;
*(u64 *)((void *)v + id) = 0;
```

is getting translated into something like 

```
/* pseudo IR */
id = ctx->protocol;
tmp = id;
tmp += -13;
if (tmp < 0xfffffff7) goto next; // <- tmp and defined bounds are discarded here
v += id;
*v = 0; // <- BPF verification fails because of unbounded variable
next:
```

which fails BPF verification.

This problem and the proposed solution were originally raised by @yonghong-song in 
https://reviews.llvm.org/D147078 
https://reviews.llvm.org/D147968 
but there were seemingly no new developments in almost a year.

This problem was also raised in discussions like
https://lore.kernel.org/all/f9c2ab1f-42a9-f5eb-f9bf-63402c6a8a62@uliege.be/T/ https://lpc.events/event/17/contributions/1648/attachments/1242/2525/Generating%20BPF%20Verifier%20Friendly%20Code%20with%20Clang.pdf

This presents a major paint point because of the absence of a clear workaround, leading to time-sinking fights against LLVM. Could some workaround be implemented until a solid solution is discussed and implemented? It may take the form of a new pragma that would disable `FoldAndOrOfICmpsUsingRanges` transformation that seems to be causing this issue.

**Comment by nikic** [comment]
If the BPF verifier does not support the range check idiom, then the BPF backend should convert it back into two comparisons.

**Comment by Spikhalskiy** [comment]
From my understanding, the main problem here is not a range check (BPF does support upper and lower bounds), but in a way how the code is transformed. 

LLVM introduces a new variable (for BPF it's a register) in the IR and performs checks on that variable. This variable is discarded and an original variable (original register) is used in the functional call later. Because LLVM performed the checks on the temporary register and discarded it, the knowledge about the boundaries is lost for BPF. While LLVM knows that it's the same value.

I will provide an example of the final BPF bytecode to demonstrate an issue. 

_It's the first problem that I investigate with LLVM. I don't have a good grasp on the project, terminology, and what pieces are responsible for what yet. If something in my explanations doesn't make sense, sorry in advance, it very likely doesn't._

**Comment by nikic** [comment]
Yes, I get what you mean :) In LLVM, `(X - C1) < C2` is the standard way to represent a two-sided range check. The BPF verifier does not understand this idiom, so the BPF backend should convert `(X - C1) < C2` back into `X > C3 && X < C4` (or `X < C3 || X > C4`).

**Comment by Spikhalskiy** [comment]
@yonghong-song I rebased your original work onto master to include & test it in our internal llvm fork and opened it as a draft PR https://github.com/llvm/llvm-project/pull/97073. Hope it may make it easier for you to return to this story when you get a chance.
